garage/tf/models/base.py

```diff
@@ -0,0 +1,297 @@
+            with _variable_scope:
+                with tf.name_scope(name=network_name):
+                    network = Network()
+                    network._inputs = inputs
+                    network._outputs = self._build(*inputs)
+        spec = self.network_output_spec()
+        if spec:
+            c = namedtuple(network_name,
+                           [*spec, 'input', 'output', 'inputs', 'outputs'])
+            if isinstance(network.outputs, tuple):
+                assert len(spec) == len(network.outputs), (
+                    'network_output_spec must have same length as outputs!')
+                self._networks[network_name] = c(
+                    *network.outputs, network.input, network.output,
+                    network.inputs, network.outputs)
+            else:
+                self._networks[network_name] = c(
+                    network.outputs, network.input, network.output,
+                    network.inputs, network.outputs)
+        else:
+            self._networks[network_name] = network
+        return network.outputs
+
+    def _build(self, *inputs):
+        """
+        Output of the model given input placeholder(s).
+
+        User should implement _build() inside their subclassed model,
+        and construct the computation graphs in this function.
+
+        Args:
+            inputs: Tensor input(s), recommended to be position arguments, e.g.
+              def build(self, state_input=None, action_input=None, name=None).
+              It would be usually same as the inputs in build().
+        Return:
+            output: Tensor output(s) of the model.
+        """
+        pass
+
+    def network_output_spec(self):
+        """
+        Network output spec.
+
+        Return:
+            *inputs: List of key(str) for the network outputs.
+        """
+        return []
+
+    @property
+    def networks(self):
+        """Networks of the model."""
+        return self._networks
+
+    @property
+    def parameters(self):
+        """Parameters of the model."""
+        return tf.get_default_session().run(self._get_variables())
+
+    @parameters.setter
+    def parameters(self, parameters):
+        """Set model parameters."""
+        variables = self._get_variables()
+        for name, var in variables.items():
+            if name in parameters:
+                var.load(parameters[name])
+            else:
+                warnings.warn('No value provided for variable {}'.format(name))
+
+    @property
+    def name(self):
+        """Name of the model."""
+        return self._name
+
+    def _get_variables(self):
+        return {v.name: v for v in tf.get_variable_scope().global_variables()}
+
+    def __getstate__(self):
+        """Object.__getstate__."""
+        new_dict = self.__dict__.copy()
+        del new_dict['_networks']
+        new_dict['_default_parameters'] = self.parameters
+        return new_dict
+
+    def __setstate__(self, dict):
```

Comment:
don't use dict as an argument, since it's a Python builtin. I'm surprised this wasn't caught by flake8.